### PR TITLE
Remove query whitespace, take 2: Function Time

### PR DESCRIPTION
**Pull request by sarahghp**
Implements approach for optionally removing removing whitespace or otherwise passing custom print functions to `HttpLink` and friends as outlined in https://github.com/apollographql/apollo-client/pull/8549.

### Checklist:

- [x] If this PR is a new feature, please reference an issue where a consensus about the design was reached (not necessary for small changes)
- [x] Make sure all of the significant new logic is covered by tests

### CODE REVIEW

**Comment by sarahghp** [comment]
@benjamn Do you mind taking a look at this draft and confirming the approach matches expectatoins? Then I'll go in and add new specs and docs.

**Comment by benjamn** [comment]
@sarahghp This is looking great/perfect so far!

If you rebase and force-push I believe the CLA check will pass. And please feel free to bump the `bundlesize` limit [here](https://github.com/sarahghp/apollo-client/blob/remove-query-whitespace-take-ii/package.json#L59) to, say, 24.8kB. That will more than cover the increase from this PR (~0.05kB, way less than before 🎉).

**Comment by sarahghp** [comment]
@benjamn @StephenBarlow This is ready for feedback on specs and docs now. Thank you!

**Review by benjamn — APPROVED** [review]
Thanks for adding tests and docs!

I rebased against `release-3.5` (since this is technically new functionality), and I expect to release these changes in `@apollo/client@3.5.0-beta.11` (the next beta release).

I'll let @StephenBarlow review these documentation changes asynchronously.

File: src/link/http/selectHttpOptionsAndBody.ts [hunk 1]
```diff
@@ -100,8 +109,11 @@ export const fallbackHttpConfig = {
   options: defaultOptions,
 };
 
+export const defaultPrinter: Printer = (ast, printer) => printer(ast);
+
 export const selectHttpOptionsAndBody = (
   operation: Operation,
+  printer: Printer,
```
Comment (SimenB) [inline src/link/http/selectHttpOptionsAndBody.ts:116]: this is a breaking change for people using `selectHttpOptionsAndBody` directly.

See https://github.com/jaydenseric/apollo-upload-client/pull/274 for an example

Comment (benjamn) [reply src/link/http/selectHttpOptionsAndBody.ts:116]: Thanks for the heads up @SimenB. I'll see what we can do to fix this.

Comment (SimenB) [reply src/link/http/selectHttpOptionsAndBody.ts:116]: Great, thanks for taking a look 👍

Comment (benjamn) [reply src/link/http/selectHttpOptionsAndBody.ts:116]: Here's what I came up with: https://github.com/apollographql/apollo-client/pull/9103

**Pull request merged**